File: src/main/java/com/tms/csp/transforms/CompoundTransformer.java
package com.tms.csp.transforms;

import com.tms.csp.ast.Exp;

import javax.annotation.Nonnull;

public class CompoundTransformer extends Transformer {

    private final Transformer[] transformers;

    public CompoundTransformer(Transformer... transformers) {
        this(null, transformers);
    }

    public CompoundTransformer(String name, Transformer... transformers) {
        this.transformers = transformers;
    }

    @Nonnull
    @Override
    public Exp transform(@Nonnull Exp in) {
        Exp exp = in;
        for (Transformer transformer : transformers) {
            exp = exp.transform(transformer);
        }
        return exp;
    }

}

File: src/main/java/com/smartsoft/csp/transforms/CompoundTransformer.kt
package com.smartsoft.csp.transforms

import com.smartsoft.csp.ast.Exp

class CompoundTransformer(name: String?, vararg transformers: Transformer) : Transformer() {

    private val transformers: Array<out Transformer>

    constructor(vararg transformers: Transformer) : this(null, *transformers) {}

    init {
        this.transformers = transformers
    }

    override fun transform(inExp: Exp): Exp {
        var exp = inExp
        for (transformer in transformers) {
            exp = exp.transform(transformer)
        }
        return exp
    }

}
